### PR TITLE
feat(HNS folders): add new resources to support HNS folders

### PR DESCRIPTION
**Pull request by gurusai-voleti**
**Release Note Template for Downstream PRs (will be copied)**

See [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) for guidance.

```release-note:new-resource
`google_storage_folder`
```

### CODE REVIEW

**Comment by kautikdk** [comment]
Please add tests for nested folders and objects inside folder.

File: mmv1/products/storage/Folder.yaml
Comment (kautikdk) [inline mmv1/products/storage/Folder.yaml:45]: This can be removed.

File: mmv1/products/storage/Folder.yaml
Comment (kautikdk) [inline mmv1/products/storage/Folder.yaml:64]: We are not planning to add this field, Right?

File: mmv1/templates/terraform/examples/storage_folder_basic.tf.tmpl
Comment (kautikdk) [inline mmv1/templates/terraform/examples/storage_folder_basic.tf.tmpl:15]: this can be removed as of now!

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (kautikdk) [inline mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl:221]: Why to list objects of the top level folder when we are trying to delete nested folder?

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (kautikdk) [inline mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl:283]: same as above comment!

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (kautikdk) [inline mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl:291]: What if there are objects inside nested folder?

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (gurusai-voleti) [reply mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl:221]: to delete folders first we need to delete all objects as non empty folders cannot be deleted, getting all the objects with top level folder prefix and delete all objects and in next step delete the sub folders and finally delete the main folder, this will clear up all the objects, subfolders and folders

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (gurusai-voleti) [reply mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl:291]: we are getting all objects with a top level folder prefix this will list all objects, here in this step only folders will remain so deleting only subfolders and folder here

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (kautikdk) [reply mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl:221]: But the scenario that I am referring is, If the folder resource is `a/b/c/d` then we try to delete it with `force_destroy` = true then as per this logic, it will try to delete everything inside folder `a/` which is not expected. we should only delete everything that is nested in `a/b/c/d` path not anything that is in upper level of path or in different path tree like `a/b/c` or `a/b/c/e/f`.

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (kautikdk) [reply mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl:291]: Ohh, I see!
I thought it won't list object that are deeply nested within sub-folders(more than 1 level). 
Got it!

File: mmv1/products/storage/Folder.yaml
Comment (gurusai-voleti) [reply mmv1/products/storage/Folder.yaml]: removed

File: mmv1/products/storage/Folder.yaml
Comment (gurusai-voleti) [reply mmv1/products/storage/Folder.yaml]: removed

File: mmv1/templates/terraform/examples/storage_folder_basic.tf.tmpl
Comment (gurusai-voleti) [reply mmv1/templates/terraform/examples/storage_folder_basic.tf.tmpl]: removed

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (gurusai-voleti) [reply mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl]: updated the logic

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (kautikdk) [inline mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl:189]: What is the need of this?

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (kautikdk) [inline mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl:195]: Not sure but if the field is not being pass down to the API request then do we need to explicitly set in updater?

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (kautikdk) [inline mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl:270]: If the folder with `force_destroy=false` that doesn't contain any sub-folder then it should get deleted without this error being thrown. 
We should do this validation after getting list of sub-folders. This was not caught in tests because there is no test for `force_destroy=false`. I recommend adding one test for the false value. 
Also It will same deletion logic for two different resources(objects and folders) so it would be good to write common logic for both and utilize that everywhere.

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (kautikdk) [inline mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl:308]: I believe we need to replace resource id from the import id using `tpgresource.ReplaceVars` function?

File: mmv1/third_party/terraform/services/storage/resource_storage_folder_test.go
Comment (kautikdk) [inline mmv1/third_party/terraform/services/storage/resource_storage_folder_test.go:55]: Just a intuition, I believe this test won't check the actual logic of force destroy deeply? The reason is we are creating all nested resources using terraform config only which will be part of terraform dependency graph. The deletion will happen by individual resource's DELETE method so there won't any object or sub-folder which will be deleted by force destroy. 
It would be good if we can some how add object in folder outside of terraform and then try to test force destroy functionality.

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (gurusai-voleti) [reply mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl:195]: not passing to API but need to manage in tf state file, customers may switch from false to true and vice versa, based on this flag we manage the deletion logic when destroying the resource

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (gurusai-voleti) [reply mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl:89]: will add false as default value

File: mmv1/third_party/terraform/services/storage/resource_storage_folder_test.go
Comment (gurusai-voleti) [reply mmv1/third_party/terraform/services/storage/resource_storage_folder_test.go:55]: objects are deleted only when force_destroy is set to true otherwise objects will not be deleted and will not  continue the delete operation

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (gurusai-voleti) [reply mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl]: updated the logic, if no subfolders are present and force_destroy set to false, allows to delete folder

File: mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl
Comment (gurusai-voleti) [reply mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl]: removed

**Review by roaks3 — COMMENTED** [review]
I just have a couple of high-level questions for context to start here, because this looks like a relatively complicated handwritten resource:
1. Is this different from managed_folder?
2. I didn't see any tickets requesting this, which would give some context if one exists, but also this is something I would have expected to get a lot of usage.
3. Could you clarify why this needs to be handwritten? It seems like the part where the folder needs to be empty would be one of the reasons.

**Comment by kautikdk** [comment]
Hi @roaks3,

> Is this different from managed_folder? 
Yes, This is different. Managed folder is specifically to set IAM on name prefixes. Managed Folder can not be treated as file structure folders. Meanwhile HNS folders are actual folders that simulates file structure hierarchy for users. Users can use these folders to organize data in various folders and sub-folders.

> I didn't see any tickets requesting this, which would give some context if one exists, but also this is something I would have expected to get a lot of usage.
@gurusai-voleti please create a GitHub ticket. Yes, This resource will have a lot of usage. 

>Could you clarify why this needs to be handwritten? It seems like the part where the folder needs to be empty would be one of the reasons.
So currently there are some complicated issues that we are seeing and that would require this resource to be hand-written. Currently this PR doesn't have that part of code. @gurusai-voleti can you please mark this PR as a draft while we are working offline?

**Comment by kautikdk** [comment]
@gurusai-voleti Can you please check if we can auto generate this resource using MMV1? Since we are not going for any complex operations, I believe we should be able to auto generate it. Not sure about path escaping for import operation, Please check if it working or not.

**Comment by gurusai-voleti** [comment]
> @gurusai-voleti Can you please check if we can auto generate this resource using MMV1? Since we are not going for any complex operations, I believe we should be able to auto generate it. Not sure about path escaping for import operation, Please check if it working or not.

DELETE and UPDATE are two actions which require custom code and will generate code using tmpl

**Comment by kautikdk** [comment]
You can remove hand-written files if MMV1 source code is enough for the feature support.

**Comment by gurusai-voleti** [comment]
> You can remove hand-written files if MMV1 source code is enough for the feature support.

Delete, import, update are customised operations so I kept the handwritten tmpl file and updated the yaml with custom_code attribute

File: mmv1/products/storage/Folder.yaml
Comment (roaks3) [inline mmv1/products/storage/Folder.yaml:71]: I think you may want some more rigorous validation here, mainly because it is used on the path. Do you know what the user will see if they provide a name like `/example_dir/`? If it results in a cryptic error, then validation might be better, and then you may want to do other things like ensure alphanumeric characters or similar.

File: mmv1/templates/terraform/custom_update/storage_folder_update.go.tmpl
Comment (roaks3) [inline mmv1/templates/terraform/custom_update/storage_folder_update.go.tmpl:7]: NIt: Since you're already adding this, you may want to go ahead and add the function that checks if this is the only field being changed (in case other fields are added later).

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (roaks3) [inline mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:17]: I would recommend moving this to the beginning, since you should be able to escape early

File: mmv1/products/storage/Folder.yaml
Comment (gurusai-voleti) [reply mmv1/products/storage/Folder.yaml:71]: if the path starts with /, the API validates the request body and rejects it as it cannot be started with /

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (gurusai-voleti) [reply mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl]: moved the logic to starting

Comment (gurusai-voleti) [reply mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:17]: we first check if the folder contains any objects, if no objects we delete the folder even force_destory set to false, if any objects are there force_destory should be set to true then we delete the objects and folders
 
if force_destroy is true non-empty folders will be deleted, if force_destory is false empty folders are deleted so first we need to check if any objects present in the folders

File: mmv1/templates/terraform/custom_update/storage_folder_update.go.tmpl
Comment (gurusai-voleti) [reply mmv1/templates/terraform/custom_update/storage_folder_update.go.tmpl:7]: for now this should be fine

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (roaks3) [reply mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:17]: Ah, thanks for clarifying, I was mixing this up with how `prevent_destroy` would be working.

File: mmv1/products/storage/Folder.yaml
Comment (roaks3) [reply mmv1/products/storage/Folder.yaml:71]: Right, but what I'm asking about is how exactly it responds when that value is rejected. For example, a 404 would be difficult for a user to understand when they actually need to remove a leading slash here.

Comment (gurusai-voleti) [reply mmv1/products/storage/Folder.yaml:71]: below is the response if folder name starts with /
"code": 409,
"message": "The parent folder does not exist.",

File: mmv1/products/storage/Folder.yaml
Comment (kautikdk) [inline mmv1/products/storage/Folder.yaml:72]: I think, We should match only trailing slash?

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (kautikdk) [inline mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:8]: Nit: Folder not bucket

File: mmv1/products/storage/Folder.yaml
Comment (gurusai-voleti) [reply mmv1/products/storage/Folder.yaml]: yes will update to only trailing /, but as Ryan has raised concern updated the regex but as per naming conventions only trailing / is fine as the info provided in another comment

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (gurusai-voleti) [reply mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl]: updated

File: mmv1/products/storage/Folder.yaml
Comment (roaks3) [inline mmv1/products/storage/Folder.yaml:52]: Nit: I would mention that items within the folder will be destroyed

File: mmv1/products/storage/Folder.yaml
Comment (roaks3) [inline mmv1/products/storage/Folder.yaml:63]: Should name be immutable, or can it be changed?

File: mmv1/products/storage/Folder.yaml
Comment (roaks3) [reply mmv1/products/storage/Folder.yaml:71]: Oh ok, that's a very permissive spec. Seems ok to leave it for now, and we can always add more client-side validation later if needed (all it would really improve is the error message, which I do think could be an issue for users thinking they need a leading slash).

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (roaks3) [inline mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:58]: Sorry if I missed this: the bucket implementation appears to attempt the bucket deletion even if there is a listError, and then only surface the error if the deletion was not successful. Was there a reason we didn't do the same here?

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (roaks3) [inline mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:60]: Nit: probably don't need this?

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (roaks3) [inline mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:84]: Nit: "... X subfolders still exist within %name, use force_destroy = true to delete all subfolders", or similar, to provide more info?

File: mmv1/templates/terraform/examples/storage_folder_basic.tf.tmpl
Comment (roaks3) [inline mmv1/templates/terraform/examples/storage_folder_basic.tf.tmpl:8]: To be more concise and clear about the functionality, would it make sense to remove `force_destroy = true` from the bucket and subfolder? I don't believe they are needed.

File: mmv1/third_party/terraform/services/storage/resource_storage_folder_test.go
Comment (roaks3) [inline mmv1/third_party/terraform/services/storage/resource_storage_folder_test.go:115]: I'm not sure I'm understanding this part. Why are we deleting `folder/` here? Shouldn't we instead do that with Terraform config (by removing it)?

**Review by roaks3 — COMMENTED** [review]
Mostly minor items. Overall code and tests look good 👍

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (gurusai-voleti) [reply mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:58]: in bucket deletion, we are not attempting to delete the bucket if the listError occurs

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (gurusai-voleti) [reply mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:60]: yes removed

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (gurusai-voleti) [reply mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:84]: yes even if subfolders exist and force_destroy set to true we can delete the sub folders  else don't delete

File: mmv1/products/storage/Folder.yaml
Comment (kautikdk) [reply mmv1/products/storage/Folder.yaml:63]: It should be immutable.
cc: @gurusai-voleti

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (kautikdk) [reply mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:58]: I think we do try to delete bucket even if the listError occurs,
[Here](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/services/storage/resource_storage_bucket.go.tmpl#L993) we are only breaking object deletion loop if error occurs. We try to delete bucket regardless of object deletion, [Reference](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/services/storage/resource_storage_bucket.go.tmpl#L1055).

File: mmv1/third_party/terraform/services/storage/resource_storage_folder_test.go
Comment (gurusai-voleti) [reply mmv1/third_party/terraform/services/storage/resource_storage_folder_test.go:115]: here we are creating a bucket and a folder with force_destroy false in the config step and later created a sub-folder inside a folder and now tried to delete parent folder, as force_destory set to false on parent folder we expect an error as parent folder is not empty (sub-folder exists),  seems this is the best way to produce this test case

File: mmv1/products/storage/Folder.yaml
Comment (gurusai-voleti) [reply mmv1/products/storage/Folder.yaml:63]: yes updated to immutable

File: mmv1/products/storage/Folder.yaml
Comment (gurusai-voleti) [reply mmv1/products/storage/Folder.yaml]: updated the description

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (gurusai-voleti) [reply mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:69]: updated logic, attempt to delete the folder after listError

File: mmv1/templates/terraform/examples/storage_folder_basic.tf.tmpl
Comment (gurusai-voleti) [reply mmv1/templates/terraform/examples/storage_folder_basic.tf.tmpl]: removed from bucket and subfolder

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (gurusai-voleti) [reply mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl]: updated message

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (roaks3) [reply mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:58]: It looks like this was added to the bucket implementation here: https://github.com/GoogleCloudPlatform/magic-modules/pull/2755. You can read the PR description, but it appears this was added to get around permissions (a user may be able to delete the bucket, but not list buckets).

I would defer to the two of you if the same issue will be present for folders.

File: mmv1/third_party/terraform/services/storage/resource_storage_folder_test.go
Comment (roaks3) [inline mmv1/third_party/terraform/services/storage/resource_storage_folder_test.go:210]: Nit: it would be nice to remove this line, as I don't think it has any influence on the behavior, and actually implies that the user would need to set `force_destroy` in all subfolders for deletion to work.

**Review by roaks3 — COMMENTED** [review]
One nit, and only open question IMO is around the "list" permission

File: mmv1/third_party/terraform/services/storage/resource_storage_folder_test.go
Comment (gurusai-voleti) [reply mmv1/third_party/terraform/services/storage/resource_storage_folder_test.go]: updated

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (gurusai-voleti) [reply mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:69]: updated the logic to attempt delete folder

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (roaks3) [inline mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:78]: Nit: we might want to still log the err if it is not nil and we get to this point. While it wouldn't change behavior, it might help when debugging, since the error is otherwise swallowed.

**Review by roaks3 — APPROVED** [review]
One nit that isn't critical, everything else LGTM

I'll also wait for a 👍 from @kautikdk before merging

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (gurusai-voleti) [reply mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:78]: added log for the err

File: mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl
Comment (kautikdk) [reply mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl:69]: Discussed offline, LGTM!

**Review by kautikdk — APPROVED** [review]
LGTM,
Please change release notes to make it more clear about new resource.

**Review by roaks3 — APPROVED** [review]
I just updated the release note for you. We actually don't include messages for new resources, just the resource name: https://googlecloudplatform.github.io/magic-modules/code-review/release-notes/

**Comment by gurusai-voleti** [comment]
> I just updated the release note for you. We actually don't include messages for new resources, just the resource name: https://googlecloudplatform.github.io/magic-modules/code-review/release-notes/

could you merge the PR if ok, I don't have permission to do a merge

**Pull request merged**
